server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -101,18 +103,21 @@ private RecoveryResponse recover(final StartRecoveryRequest request) throws IOEx
         RecoverySourceHandler handler = ongoingRecoveries.addNewRecovery(request, shard);
         logger.trace("[{}][{}] starting recovery to {}", request.shardId().getIndex().getName(), request.shardId().id(),
             request.targetNode());
+        Runnable removeRecoveryHandler = () -> ongoingRecoveries.remove(shard, handler);
         try {
-            return handler.recoverToTarget();
+            handler.recoverToTarget(ActionListener.runAfter(listener, removeRecoveryHandler));
```

Comment:
my problem with this is that once I passed the listener I expect it to be called. so if we fail here with an exception and the listener is not invoked that's a bug.